examples/ed25519/src/ed25519.rs

```diff
@@ -0,0 +1,909 @@
+        VerifyResult::Ok(())
+    } else {
+        VerifyResult::Err(Error::InvalidSignature)
+    }
+}
+
+/// Cofactorless verification.
+/// Rejects non-canonical encoding of points.
+/// Allows small-order points.
+pub fn ietf_cofactorless_verify(
+    pk: PublicKey,
+    signature: Signature,
+    msg: &ByteSeq,
+) -> VerifyResult {
+    let b = decompress(BASE).unwrap();
+    let a = decompress(pk).ok_or(Error::InvalidPublickey)?;
+    let r_bytes = CompressedEdPoint::from_slice(&signature, 0, 32);
+    let s_bytes = SerializedScalar::from_slice(&signature, 32, 32);
+    if !check_canonical_scalar(s_bytes) {
+        VerifyResult::Err(Error::InvalidS)?;
+    }
+    let r = decompress(r_bytes).ok_or(Error::InvalidR)?;
+    let s = Scalar::from_byte_seq_le(s_bytes);
+    let k = scalar_from_hash(sha512(&r_bytes.concat(&pk).concat(msg)));
+
+    let sb = point_mul(s, b);
+    let ka = point_mul(k, a);
+
+    if point_eq(sb, point_add(r, ka)) {
+        VerifyResult::Ok(())
+    } else {
+        VerifyResult::Err(Error::InvalidSignature)
+    }
+}
+
+fn is_identity(p: EdPoint) -> bool {
+    point_eq(p, point_identity())
+}
+
+/// Algorithm 2 from <https://eprint.iacr.org/2020/1244.pdf>.
+/// Cofactored verification.
+/// Rejects non-canonical encoding of points.
+/// Rejects small-order points for the public key.
+pub fn alg2_verify(pk: PublicKey, signature: Signature, msg: &ByteSeq) -> VerifyResult {
+    let b = decompress(BASE).unwrap();
+    let a = decompress(pk).ok_or(Error::InvalidPublickey)?;
+    if is_identity(point_mul_by_cofactor(a)) {
+        VerifyResult::Err(Error::SmallOrderPoint)?;
+    }
+    let r_bytes = CompressedEdPoint::from_slice(&signature, 0, 32);
+    let s_bytes = SerializedScalar::from_slice(&signature, 32, 32);
+    if !check_canonical_scalar(s_bytes) {
+        VerifyResult::Err(Error::InvalidS)?;
+    }
+    let r = decompress(r_bytes).ok_or(Error::InvalidR)?;
+    let s = Scalar::from_byte_seq_le(s_bytes);
+    let k = scalar_from_hash(sha512(&r_bytes.concat(&pk).concat(msg)));
+
+    let sb = point_mul_by_cofactor(point_mul(s, b));
+    let rc = point_mul_by_cofactor(r);
+    let ka = point_mul_by_cofactor(point_mul(k, a));
+
+    if point_eq(sb, point_add(rc, ka)) {
+        VerifyResult::Ok(())
+    } else {
+        VerifyResult::Err(Error::InvalidSignature)
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct BatchEntry(PublicKey, ByteSeq, Signature);
+
+/// Batch verification.
+/// Cofactored verification.
+/// Allows non-canonical encoding of points.
+/// Allows small-order points.
+pub fn zcash_batch_verify(entries: Seq<BatchEntry>, entropy: &ByteSeq) -> VerifyResult {
+    if entropy.len() < 16 * entries.len() {
+        VerifyResult::Err(Error::NotEnoughRandomness)?;
+    }
+    let mut s_sum = Scalar::ZERO();
+    let mut r_sum = point_identity();
+    let mut a_sum = point_identity();
+    for i in 0..entries.len() {
+        let BatchEntry(pk, msg, signature) = entries[i].clone();
```

Comment:
Since you're cloning the values anyway you could have `entries` be a reference. There's no need to take ownership of this.